### PR TITLE
fix: allow URL object to be passed to local `fetch()` calls

### LINKED ISSUE #1588: 🐛 BUG: fetch() does not accept URL objects when testing locally but does when deployed
### What version of `Wrangler` are you using?

2.0.23

### What operating system are you using?

Windows

### Describe the Bug

1. Clone the example repo: https://github.com/ardislu/cloudflare-pages-url-test
2. Run `npx wrangler pages dev .`
3. Go to `http://localhost:8788/api` and see it returns an `Invalid URL` errror
4. Deploy the repo to Cloudflare Pages and see that it's now working as expected. I deployed my example repo to: https://cloudflare-pages-url-test.pages.dev/

I figured out a workaround by manually calling `.toString()` on the URL object before passing it to `fetch`:
```javascript
const exampleUrl = new URL('https://example.com');

await fetch(exampleUrl); // Throws an "Invalid URL" error when testing locally but not when deployed
await fetch(exampleUrl.toString()); // Works both locally and when deployed
```

### PR DESCRIPTION
**Pull request by dario-piotrowicz**
Due to the local fetch check, URL objects are wrongly not accepted by `fetch()` calls,
amend the `checkedFetch` function in order to allow such scenario (which
anyways works after deployment).

resolves #1588

### CODE REVIEW

**Comment by dario-piotrowicz** [comment]
Hi there :slightly_smiling_face: , I'm just trying to do this small bug fix, I hope my change makes sense :slightly_smiling_face: 

Anyways the type of the request is still `Request | string`, so although this works fine typescript is not happy :cry: I had a look into [workers-types](https://github.com/cloudflare/workers-types) but if I understood correctly the `fetch` param types there are all auto-generated?
do the `fetch` param types need to be updated in the workers runtime? (PS: I had a look just to be sure and according to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/fetch#parameters) `fetch` is indeed supposed to accept a `URL` object)

**Comment by dario-piotrowicz** [comment]
Also I couldn't fine unit tests regarding this `fetch` patching (that's why I didn't add any), is it not being tested? or am I just missing the tests?

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Thanks for putting this together @dario-piotrowicz.
From looking at the MDN docs for `fetch()`, it says:

> The fetch() method's parameters are identical to those of the [Request()](https://developer.mozilla.org/en-US/docs/Web/API/Request/Request) constructor.

So I would propose a simpler fix which is to always pass the fetch params to a `Request()` constructor and get the URL that way. E.g.

```ts
checkedFetch(request, init) {
  const url = new URL((new Request(request, init)).url;
  ...
}
```

WDYT?

And sorry that I never made any tests for this 😱 I don't know why...
It is a little awkward mocking out the real `fetch()` but we do it elsewhere.

**Comment by dario-piotrowicz** [comment]
> Thanks for putting this together @dario-piotrowicz. From looking at the MDN docs for `fetch()`, it says:
> 
> > The fetch() method's parameters are identical to those of the [Request()](https://developer.mozilla.org/en-US/docs/Web/API/Request/Request) constructor.
> 
> So I would propose a simpler fix which is to always pass the fetch params to a `Request()` constructor and get the URL that way. E.g.
> 
> ```ts
> checkedFetch(request, init) {
>   const url = new URL((new Request(request, init)).url;
>   ...
> }
> ```
> 
> WDYT?
> 
> And sorry that I never made any tests for this scream I don't know why... It is a little awkward mocking out the real `fetch()` but we do it elsewhere.

Sounds good thanks :smiley: 

About the tests if you want I can have a try in adding some :smile: 

by the way, what about the types then? how can I get the param types updated?

**Comment by petebacondarwin** [comment]
Regarding the typings, I think we probably could start by creating an issue here: https://github.com/cloudflare/workers-types.
From the docs, the request can be anything with a "stringifier" which doesn't seem very easy to define in pure TS.
@mrbbot could confirm but I think the fix would be in the Workers Runtime code internally at Cloudflare.

**Comment by dario-piotrowicz** [comment]
I see, thanks @petebacondarwin and @mrbbot :slightly_smiling_face: , shall I then proceed and create an issue in https://github.com/cloudflare/workers-types?

**Comment by petebacondarwin** [comment]
Yes please!

**Review by petebacondarwin — APPROVED** [review]

File: packages/wrangler/templates/checked-fetch.js [hunk 1]
```diff
@@ -1,9 +1,7 @@
 const urls = new Set();
 
 export function checkedFetch(request, init) {
-	const url = new URL(
-		(typeof request === "string" ? new Request(request, init) : request).url
-	);
+	const url = new URL(new Request(request, init).url);
```
Comment (GregBrimble) [inline packages/wrangler/templates/checked-fetch.js:4]: Ah, I think I got bitten by this change just the other day. Something to do with the request body being consumed twice. I'll look into it, but this should maybe be reverted.

Comment (petebacondarwin) [reply packages/wrangler/templates/checked-fetch.js:4]: Yes let's revert this. I can see this error using the reproduction in the linked issue.

https://github.com/cloudflare/wrangler2/issues/2562#issuecomment-1400855328

Comment (petebacondarwin) [reply packages/wrangler/templates/checked-fetch.js:4]: Here: https://github.com/cloudflare/wrangler2/pull/2597

Comment (dario-piotrowicz) [reply packages/wrangler/templates/checked-fetch.js:4]: sorry about the bug :cry: (thanks for the fix Pete :bow:)

Comment (petebacondarwin) [reply packages/wrangler/templates/checked-fetch.js:4]: I never thought to test it on POST requests coming out of a Worker.
The reproduction required a `fetch()` call to a `Request` that has a body (but GET requests cannot have bodies).

**Pull request merged**